### ISSUE TITLE
--config-setting example

### ISSUE THREAD

**Issue by abitrolly** [pypa/build#205]
It doesn't seem that -C option has any effect.
```shell
✗ python -m build --no-isolation -C --no-cache
usage: python -m build [-h] [--version] [--sdist] [--wheel] [--outdir dir]
                       [--skip-dependencies] [--no-isolation]
                       [--config-setting CONFIG_SETTING]
                       [srcdir]
python -m build: error: argument --config-setting/-C: expected one argument
✗ python -m build --no-isolation -C '--no-cache'
usage: python -m build [-h] [--version] [--sdist] [--wheel] [--outdir dir]
                       [--skip-dependencies] [--no-isolation]
                       [--config-setting CONFIG_SETTING]
                       [srcdir]
python -m build: error: argument --config-setting/-C: expected one argument
```

**Comment by abitrolly** [comment]
`setuptools` doesn't fully support `config-settings` - https://github.com/pypa/setuptools/issues/2491

That's why an example is needed.

Maybe document which backends are compatible, and which are not?
Maybe add an interface for backends to report settings that they support?

**Comment by uranusjr** [comment]
How should the example be structured? All the frontend can do is to `--config-settings` directly to the backend, and an example in `build` would be something along the line *you can pass stuff with this flag, but what they are and how they work depends on the build backend*, which not really achieve anything. You should find a backend that actually supports this flag first to make the example worthwhile.

**Comment by layday** [comment]
> It doesn't seem that -C option has any effect.

The value is being interpreted as an argument to `build` because it begins with a hyphen.  In such case you must either use an equals sign or you must not leave a space between `-C` and the configuration value:

```sh
python -m build --no-isolation -C--no-cache
```

**Comment by abitrolly** [comment]
> How should the example be structured?

There is a story of the user who tried to pass `--no-cache` flag and `--help` flag to `setuptools` and didn't get any response. Following the story gets the structure - from trying to figure out how to pass the option to debugging if it was sent, and revealing that backend is not processing it. Then running the backend manually.

**Comment by abitrolly** [comment]
> The value is being interpreted as an argument to build because it begins with a hyphen.

Is it how the argument parsing is supposed to work if the setting is defined with non-optional value as `-C CONFIG_SETTING`?

**Comment by layday** [comment]
No, it's a well-known issue with `argparse`.

**Comment by gaborbernat** [comment]
> No, it's a well-known issue with `argparse`.

Do you have an upstream ticket for this, if it's well-known. And considering this limitation I suppose this ticket may be addressed by documenting your comments from https://github.com/pypa/build/issues/205#issuecomment-760907314.

**Comment by layday** [comment]
https://bugs.python.org/issue9334

**Comment by FFY00** [comment]
I agree this is an issue in argparse and should be addressed there. However, there are still two alternative forms, `-C=--no-cache` and `-C--no-cache`. I would recommend the `-C--no-cache` format. Since this is not a very big issue, I struggle to justify moving away from argparse. We should add an examples page to the documentation to mitigate this and avoid more people running into the issue and having to go search for a solution by themselves.

**Comment by abitrolly** [comment]
It doesn't seem that 10 years old bug will gain traction any time soon. Is switching to patched argparse possible?

**Comment by abitrolly** [comment]
> We should add an examples page to the documentation to mitigate this and avoid more people running into the issue and having to go search for a solution by themselves.

The output of `--help` should be changed.

**Comment by FFY00** [comment]
Well, since this issue is now in my interest, it is in my TODO, but I am not sure when I will be able to get to it, or how long it would take to get a patch reviewed, merged and released. Moving to a patched argparse would mean adding a new dependency, which I am very strongly opposed to as that would make the bootstrapping process more difficult, as it if were already not difficult enough :upside_down_face:. The most sensible solution to this would probably hack around this by manually joining `-C --something` to `-C--something` before passing the arguments to argparse, but I am not very happy with that. If this starts crippling down many users, I will consider.

I am sorry for the trouble this is causing, but there are alternative forms you can use, so this issue shouldn't be a blocker in any way. Instead of doing `-C --no-cache`, do `-C--no-cache`, which I think is a far better form anyway.

**Comment by gaborbernat** [comment]
Until argparse gets fixed upstream I'm happy that we document this behavior in the help message.

**Comment by gaborbernat** [comment]
Fixed by #207. Thanks!
